app/models/vocabulary.rb

```diff
@@ -1,15 +1,77 @@
 class Vocabulary < ApplicationRecord
+  ALPHANUMERIC_UNDERSCORE_KEY_REGEX = /\A[a-z]+[a-z0-9_]*\z/
+  RESERVED_FIELD_NAMES = %w(
+    pref_label alt_label uri uri_hash vocabulary vocabulary_id,
+    authority term_type custom_fields uuid created_at updated_at
+  ).freeze
+  DATA_TYPES = %w(string number boolean).freeze
+
   has_many :terms, dependent: :destroy
 
   validates :string_key, presence: true, uniqueness: true, format: {
-    with: /\A[a-z]+[a-z0-9_]*\z/,
+    with: ALPHANUMERIC_UNDERSCORE_KEY_REGEX,
     message: 'only allows lowercase alphanumeric characters and underscores'
   }
   validates :label,      presence: true
+  validate :validate_custom_fields
 
   store :custom_fields, coder: JSON
 
   def to_api
     as_json(only: [:string_key, :label, :custom_fields])
   end
+
+  def add_custom_field(options = {})
```

Comment:
Why set a default param value of `{}` if that's never valid?

Reply:
I need to have an empty hash in order to generate the first error, because if there isn't a hash i'll get an error about trying to run a method on nil.

Reply:
I feel like we should instead have a check for nil `options` in the method.  The way that the method signature currently looks seems to suggest that it actually is valid to pass no args because the parameter fallback will take care of things.

Reply:
I am not sure how adding the check in the method would make the method signature clearer. Can you psuedo code what you mean?

Reply:
I mean that as a person glancing through the ruby API for this class, having `options = {}` as one of the params suggests to me that there's a valid case for not supplying any params for this method, but the internals of the method prove that isn't the case.

Reply:
You can keep things as-is if you'd prefer.  This is just a suggestion.  Either way, I'll approve the PR now!